Refactor :: oauth middleware

srrathi opened the pull request:
## Pull Request Template

**Description:**

Refactored oauth middleware  by breaking it into smaller functions for reducing it's cognitive complexity.
Solves #997 

**Checklist:**

-   [x] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [x] All new code is covered by unit tests.
-   [x] This PR does not decrease the overall code coverage.
-   [x] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

vipul-rawat commented on pkg/gofr/http/middleware/oauth.go:
```diff
@@ -152,6 +133,39 @@ func OAuth(key PublicKeyProvider) func(inner http.Handler) http.Handler {
+// ValidateAuthorizationHeader validates the format and presence of the Authorization header.
+func validateAuthorizationHeader(authHeader string) error {
+	if authHeader == "" {
+		return errors.New("authorization header is required")
```
```suggestion
		return errors.New("Authorization header is required")
```

ccoVeille commented on pkg/gofr/http/middleware/oauth.go:
```diff
@@ -152,6 +133,39 @@ func OAuth(key PublicKeyProvider) func(inner http.Handler) http.Handler {
 }
 
+// ValidateAuthorizationHeader validates the format and presence of the Authorization header.
+func validateAuthorizationHeader(authHeader string) error {
+	if authHeader == "" {
+		return errors.New("authorization header is required")
+	}
+
+	headerParts := strings.Split(authHeader, " ")
+	if len(headerParts) != 2 || headerParts[0] != "Bearer" {
+		return errors.New("authorization header format must be Bearer {token}")
+	}
+
+	return nil
+}
+
+// ExtractTokenFromHeader extracts the JWT token from the Authorization header.
+func extractTokenFromHeader(authHeader string) string {
+	headerParts := strings.Split(authHeader, " ")
+	return headerParts[1]
+}
```
This 2 methods could be merged

ccoVeille commented on pkg/gofr/http/middleware/oauth.go:
```diff
@@ -152,6 +133,39 @@ func OAuth(key PublicKeyProvider) func(inner http.Handler) http.Handler {
+		return errors.New("authorization header is required")
+	}
+
+	headerParts := strings.Split(authHeader, " ")
+	if len(headerParts) != 2 || headerParts[0] != "Bearer" {
```
I would use strings.CutPrefix(authHeader, "Bearer ") then check the found flag

srrathi commented:
@vipul-rawat @ccoVeille Thanks for the review. I have added the requested changes.

aryanmehrotra reviewed (changes requested):
@srrathi your tests are breaking in the local system as well. 
Probably something is wrong with the implementation, could you verify

srrathi commented:
> @srrathi your tests are breaking in the local system as well. Probably something is wrong with the implementation, could you verify

@aryanmehrotra Thanks for informing. These test cases are breaking
```
=== RUN   TestOAuthInvalidTokenFormat
    oauth_test.go:68: 
        	Error Trace:	/home/runner/work/gofr/gofr/pkg/gofr/http/middleware/oauth_test.go:68
        	Error:      	"authorization header format must be ***" does not contain "Authorization header format must be ***"
        	Test:       	TestOAuthInvalidTokenFormat
--- FAIL: TestOAuthInvalidTokenFormat (0.00s)
=== RUN   TestOAuthEmptyAuthHeader
    oauth_test.go:92: 
        	Error Trace:	/home/runner/work/gofr/gofr/pkg/gofr/http/middleware/oauth_test.go:92
        	Error:      	"authorization header is required\n" does not contain "Authorization header is required"
        	Test:       	TestOAuthEmptyAuthHeader
--- FAIL: TestOAuthEmptyAuthHeader (0.01s)
```

**Reason** - because I changed the error message being started from a capital letter as it is causing linting error and golang error messages should start from lower case letter. I have changed the same in the test cases as well so it should fix it.

aryanmehrotra commented:
@srrathi pipeline is still failing could you please look into it.

srrathi commented:
> @srrathi pipeline is still failing could you please look into it.

@aryanmehrotra This is what failing the pipeline, no idea why but will check in the weekend.

```
=== RUN   Test_SQLRetryConnectionInfoLog
could not connect with 'user' user to 'test' database at 'host:3201', error: dial tcp: lookup host on 127.0.0.53:53: server misbehaving
    sql.go:200: Unexpected call to *sql.MockMetrics.SetGauge([app_sql_open_connections 1]) at /home/runner/work/gofr/gofr/pkg/gofr/datasource/sql/sql.go:200 because: 
        expected call at /home/runner/work/gofr/gofr/pkg/gofr/datasource/sql/sql_test.go:275 doesn't match the argument at index 1.
        Got: 1 (float64)
        Want: is equal to 0 (float64)
        expected call at /home/runner/work/gofr/gofr/pkg/gofr/datasource/sql/sql_test.go:276 doesn't match the argument at index 0.
        Got: app_sql_open_connections (string)
        Want: is equal to app_sql_inUse_connections (string)
    controller.go:98: missing call(s) to *sql.MockMetrics.SetGauge(is equal to app_sql_open_connections (string), is equal to 0 (float64)) /home/runner/work/gofr/gofr/pkg/gofr/datasource/sql/sql_test.go:275
    controller.go:98: missing call(s) to *sql.MockMetrics.SetGauge(is equal to app_sql_inUse_connections (string), is equal to 0 (float64)) /home/runner/work/gofr/gofr/pkg/gofr/datasource/sql/sql_test.go:276
    controller.go:98: aborting test due to missing call(s)
--- FAIL: Test_SQLRetryConnectionInfoLog (0.12s)
FAIL
coverage: 3.4% of statements in gofr.dev/pkg/...
FAIL	gofr.dev/pkg/gofr/datasource/sql	0.249s

```
